src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java

```diff
@@ -0,0 +1,432 @@
+        Utils.nonNull(positions);
+        Utils.nonNull(model);
+        final List<D> dataList = Collections.unmodifiableList(new ArrayList<>(data));
+        final List<T> positionList = Collections.unmodifiableList(new ArrayList<>(positions));
+        if (dataList.size() != positionList.size()) {
+            throw new IllegalArgumentException("the data sequence and position sequence must have the same number of elements");
+        }
+        final List<S> states = model.hiddenStates();
+        Utils.nonNull(states, "the model hidden state list cannot be null");
+        if (states.size() == 0) {
+            throw new IllegalArgumentException("the model hidden state list must contain at least one element");
+        }
+
+        final double[] forwardProbabilities = calculateLogForwardProbabilities(model, dataList, positionList);
+        final double[] backwardProbabilities = calculateLogBackwardProbabilities(model, dataList, positionList);
+
+        return new ArrayResult<>(dataList, positionList, model, forwardProbabilities, backwardProbabilities);
+    }
+
+    /**
+     * Calculates the backward probabilities (the backward phase of the algorithm). These are in log scale.
+     * <p>
+     *     The returned array is uni-dimensional. Back probabilities are arranged by position, thus the
+     *     back probability for the ith hidden state and jth position is stored in the
+     *     index: i + j * number-of-hidden-states.
+     * </p>
+     * @param model the HMM model.
+     * @param dataList the observed data sequence.
+     * @param positionList the observation time/position points.
+     * @param <D> the observed data type.
+     * @param <T> the observation time/position type.
+     * @param <S> the hidden state type.
+     * @return never {@code null}, the calling code can modify this array at will.
+     * @throws IllegalArgumentException if the {@code model} does not recognize any of the values in {@code dataList}
+     *   or {@code positionList}.
+     */
+    private static <D, T, S> double[] calculateLogBackwardProbabilities(final HiddenMarkovModel<D, T, S> model,
+                                                                   final List<D> dataList,
+                                                                   final List<T> positionList) {
+
+        final List<S> states = model.hiddenStates();
+        final int width = states.size();
+        final int length = dataList.size();
+        final double[] result = new double[width * length];
+
+        // Empty data? then we just return the empty array:
+        if (length == 0) {
+            return result;
+        }
+
+        // We first initialize the probabilities at the first position.
+        // result already contains appropriate values for the backward-probabilities
+        // of the last state = 1, (0 in log scale).
+
+        final double[] logSumBuffer = new double[states.size()];
+        for (int t = length - 2; t >= 0; --t) {
```

Comment:
`t` and `u` need clearer names, like `thisPosition` and `nextPosition` or `here` and `next` etc.